### DISCUSSION TITLE
[Q] Make OPAM recognize an existing version (manually installed)

### DISCUSSION THREAD

**Discussion by anzhi0708** [ocaml/opam#5483] [Q&A] [answered]
After compiling OCaml 5.1 from source, OPAM still recognizes **only** previously installed switch

```
~ $ opam switch list
#  switch  compiler
          description
→  4.14.1  ocaml-base-compiler.4.14.1
          4.14.1
```

I compiled & installed OCaml 5 to `bin/` tho

```
~ $ ocaml --version
The OCaml toplevel, version 4.14.1
~ $ $PREFIX/bin/ocaml --version
The OCaml toplevel, version 5.1.0+dev1-2022-06-09
```
I want version 5.1 listed in `opam switch list` too.

These two versions have different installation directories
```
~ $ which ocaml
/data/data/com.termux/files/home/.opam/4.14.1/bin/ocaml
~ $ which $PREFIX/bin/ocaml
/data/data/com.termux/files/usr/bin/ocaml
```

Send help :3

(Summary:
- version 5.1 was NOT installed by OPAM
- version 4 WAS installed by OPAM
- Now OPAM does not realize the existence of version 5
)

**Comment by kit-ty-kate** [comment, marked as answer]
The problem is not in opam but in opam-repository which lacks the necessary `ocaml-system.5.1.0` package.

So far we've never had this package during the development portion because the need didn't really arose as people usually use the builtin `ocaml-variants.5.1.0+trunk`.

For you though, there is two ways to fix that:
* Add the missing `ocaml-system` package in opam-repository (but I'm not sure what to use as value to match for the `sys-ocaml-version` variable so it might be a bit tricky)
* Follow the OCaml compiler documentation: https://github.com/ocaml/ocaml/blob/trunk/HACKING.adoc#testing-with-opam and reinstall it in a switch this way.

I've opened an issue in https://github.com/ocaml/opam-repository/issues/23515
I'll close this discussion as I see you've already opened https://github.com/ocaml/ocaml/discussions/12110 for the second solution and you can comment on the new issue if you want to add something for the first solution.

**Reply by anzhi0708** [reply]
Thank you very much!

**Comment by rjbou** [comment]
fyi, [`ocaml-system`](https://github.com/AltGr/ocaml-system) is a package that you can pin to install custom system packages (that are not in default repository).
